vro-streamlit/src/vro_streamlit/auth/auth_service.py

```diff
@@ -1,9 +1,78 @@
-from vro_streamlit.auth.user import User
+import time
 
+import requests
 
-def log_in() -> User:
-    return User('test')
+from vro_streamlit.auth.response_models import DeviceFlowResponse, UserInfoResponse
+from vro_streamlit.config import GITHUB_CLIENT_ID, GITHUB_CLIENT_SECRET
 
+# Replace with your actual values
+CLIENT_ID = GITHUB_CLIENT_ID
+CLIENT_SECRET = GITHUB_CLIENT_SECRET
 
-def log_out() -> bool:
-    return True
+DEVICE_CODE_URL = 'https://github.com/login/device/code'
+TOKEN_URL = 'https://github.com/login/oauth/access_token'
+USER_URL = 'https://api.github.com/user'
+REVOKE_URL = f'https://api.github.com/applications/{CLIENT_ID}/token'
+
+
+def initiate_device_flow() -> DeviceFlowResponse:
+    """Initiate the OAuth Device Flow and return a DeviceFlowResponse object."""
+    response = requests.post(
+        DEVICE_CODE_URL,
+        data={'client_id': CLIENT_ID, 'scope': 'read:user, read:org'},
+        headers={'Accept': 'application/json'},
+    )
+    response.raise_for_status()
+    data = response.json()
+    return DeviceFlowResponse(
+        data['device_code'],
+        data['user_code'],
+        data['verification_uri'],
+        int(data['expires_in']),
+        int(data['interval']),
+    )
+
+
+def poll_for_token(device_code: str, interval: int) -> str:
+    """Poll GitHub for an access token until the user authorizes the app."""
+    while True:
+        response = requests.post(
+            TOKEN_URL,
+            data={
+                'client_id': CLIENT_ID,
+                'device_code': device_code,
+                'grant_type': 'urn:ietf:params:oauth:grant-type:device_code',
+            },
+            headers={'Accept': 'application/json'},
+        )
+
+        data = response.json()
+        if 'access_token' in data:
+            return str(data['access_token'])
+        elif 'error' in data and data['error'] != 'authorization_pending':
+            raise Exception(f"Error: {data['error_description']}")
+
+        time.sleep(interval)
+
+
+def fetch_user_info(access_token: str) -> UserInfoResponse:
+    """Fetch user info using the access token."""
+    response = requests.get(
+        USER_URL,
+        headers={'Authorization': f'token {access_token}'},
+    )
+    response.raise_for_status()
+    data = response.json()
+    return UserInfoResponse(**data)
+
+
+def log_out(access_token: str) -> bool:
+    """Revoke the access token."""
+    response = requests.delete(
+        REVOKE_URL,
+        auth=(CLIENT_ID, CLIENT_SECRET),  # Basic Authentication
+        json={'access_token': access_token},  # Send token in the request body
+        headers={'Accept': 'application/vnd.github+json'},  # Recommended header for GitHub API
+    )
+    response.raise_for_status()
+    return response.status_code == 204
```

Comment:
Do we need to handle any exceptions raised by `response.raise_for_status()`? E.g.:
```python
def log_out(access_token: str) -> bool:
    try:
        response = requests.delete(
            REVOKE_URL,
            auth=(CLIENT_ID, CLIENT_SECRET),
            json={"access_token": access_token},
            headers={"Accept": "application/vnd.github+json"},
        )
        response.raise_for_status()
        return response.status_code == 204
    except requests.RequestException as e:
        logging.error(f"Failed to revoke access token. Error: {e}")
        return False
```